evm/src/test/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessorTest.java

```diff
@@ -41,59 +41,35 @@
 import org.apache.tuweni.bytes.Bytes;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
-import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.ValueSource;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 
 @ExtendWith(MockitoExtension.class)
 abstract class AbstractMessageProcessorTest<T extends AbstractMessageProcessor> {
-
   @Mock MessageFrame messageFrame;
   @Mock OperationTracer operationTracer;
   @Mock Deque<MessageFrame> messageFrameStack;
-  @Mock WorldUpdater worldUpdater;
 
   protected abstract T getAbstractMessageProcessor();
 
-  @ParameterizedTest
-  @ValueSource(ints = {0, 1})
-  void shouldNotTraceContextIfStackSizeIsZero(final int stackSize) {
-    when(messageFrame.getMessageStackSize()).thenReturn(stackSize);
-    when(messageFrame.getState())
-        .thenReturn(MessageFrame.State.COMPLETED_SUCCESS, MessageFrame.State.COMPLETED_FAILED);
-    when(messageFrame.getMessageFrameStack()).thenReturn(messageFrameStack);
-
-    getAbstractMessageProcessor().process(messageFrame, operationTracer);
-
-    verify(operationTracer, never()).traceContextEnter(messageFrame);
-    verify(operationTracer, never()).traceContextExit(messageFrame);
-  }
-
-  @ParameterizedTest
-  @ValueSource(ints = {2, 3, 5, 15, Integer.MAX_VALUE})
-  void shouldTraceContextIfStackSizeIsGreaterZeroAndSuccess(final int stackSize) {
-    when(messageFrame.getMessageStackSize()).thenReturn(stackSize);
+  @Test
+  void shouldTraceExitForContext() {
+    when(messageFrame.getWorldUpdater()).thenReturn(new ToyWorld());
     when(messageFrame.getState()).thenReturn(MessageFrame.State.COMPLETED_SUCCESS);
     when(messageFrame.getMessageFrameStack()).thenReturn(messageFrameStack);
-    when(messageFrame.getWorldUpdater()).thenReturn(worldUpdater);
 
     getAbstractMessageProcessor().process(messageFrame, operationTracer);
 
-    verify(operationTracer, times(1)).traceContextEnter(messageFrame);
-    verify(operationTracer, times(1)).traceContextExit(messageFrame);
+    verify(operationTracer).traceContextExit(messageFrame);
   }
 
-  @ParameterizedTest
-  @ValueSource(ints = {2, 3, 5, 15, Integer.MAX_VALUE})
-  void shouldTraceContextIfStackSizeIsGreaterZeroAndFailure(final int stackSize) {
-    when(messageFrame.getMessageStackSize()).thenReturn(stackSize);
+  @Test
+  void shouldTraceExitEvenIfContextFailed() {
     when(messageFrame.getState()).thenReturn(MessageFrame.State.COMPLETED_FAILED);
     when(messageFrame.getMessageFrameStack()).thenReturn(messageFrameStack);
 
     getAbstractMessageProcessor().process(messageFrame, operationTracer);
 
-    verify(operationTracer, times(1)).traceContextEnter(messageFrame);
```

Comment:
shouldn't traceContextEnter still be called?

Reply:
No, because the mocker makes the frame states directly jump to `COMPLETED_FAILED`.

Testing `traceContextEnter` here would required to mock enough of the `MessageFrame` to run the whole `process` on it.

Reply:
I think add a comment to the test to explain why contextEnter is not expected